infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -357,26 +357,45 @@ private void onLeadershipAcquire() {
         }
 
         sessionManager.notifyLeadershipChange();
-        sessionManager.refresh(topologyDescriptor);
-        // check if all the sessions in system tables are valid
-        sessionManager.removeStaleSessionOnLeadershipAcquire();
+        performReplicationSetup(topologyDescriptor, true);
+        // record metrics about the lock acquired.
+        lockAcquireSample = recordLockAcquire();
+        processCountOnLockAcquire();
+    }
 
+    private void performReplicationSetup(TopologyDescriptor topology, boolean onLeadershipAcquire) {
+        sessionManager.refresh(topology);
 
-        // If local cluster is Source, start the client config listener.
-        if (isSource(topologyDescriptor)) {
-            sessionManager.startClientConfigListener();
+        // If not invoked as part of leadership change,
+        // If invoked as part of leadership acquisition, check if all the sessions in system tables are valid
+        if (onLeadershipAcquire) {
+            sessionManager.removeStaleSessionOnLeadershipAcquire();
+        } else {
+            // Otherwise, this method was invoked due to a topology change.  So update the topology descriptor
+            topologyDescriptor = topology;
         }
 
-        setupConnectionComponents();
+        // Setup the connection components if this node is the leader
+        if (sessionManager.getReplicationContext().getIsLeader().get()) {
+            setupConnectionComponents();
+        }
 
-        if (isSource(topologyDescriptor)) {
-            logReplicationEventListener = new LogReplicationEventListener(this, runtime);
-            logReplicationEventListener.start();
+        // If invoked on a topology change and no longer a Source, stop the listeners
+        if (!onLeadershipAcquire && !isSource(topologyDescriptor)) {
+            if (logReplicationEventListener != null) {
+                logReplicationEventListener.stop();
+            }
+            sessionManager.stopClientConfigListener();
         }
 
-        // record metrics about the lock acquired.
-        lockAcquireSample = recordLockAcquire();
-        processCountOnLockAcquire();
+        // If the current cluster is a Source and a leader, start the listeners
+        if (isSource(topologyDescriptor) && sessionManager.getReplicationContext().getIsLeader().get()) {
+            if (logReplicationEventListener == null) {
+                logReplicationEventListener = new LogReplicationEventListener(this, runtime);
+            }
+            logReplicationEventListener.start();
+            sessionManager.startClientConfigListener();
```

Comment:
I remember we discussed this before: on master the event listener was started on all the nodes, and the listener itself checks if it is leader node or not and then drops messages accordingly. Since we now start it on leader node only, maybe it's a good time to do the change there as well? 
https://github.com/CorfuDB/CorfuDB/blob/a57985d74b6b0ef1dddc4a14dd0c75ff888dcd32/infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LogReplicationEventListener.java#L61-L66

Reply:
+1

Reply:
+1